typing/typecore.ml

```diff
@@ -4077,43 +4078,45 @@ and type_let ?(check = fun s -> Warnings.Unused_var s)
        are unused. If this is the case, for local declarations, the issued
        warning is 26, not 27.
      *)
-    List.map
-      (fun pat ->
-        if not warn_unused then pat, None
-        else
-          let some_used = ref false in
-            (* has one of the identifier of this pattern been used? *)
-          let slot = ref [] in
-          List.iter
-            (fun id ->
-              let vd = Env.find_value (Path.Pident id) new_env in
-              (* note: Env.find_value does not trigger the value_used event *)
-              let name = Ident.name id in
-              let used = ref false in
-              if not (name = "" || name.[0] = '_' || name.[0] = '#') then
-                add_delayed_check
-                  (fun () ->
-                    if not !used then
-                      Location.prerr_warning vd.Types.val_loc
-                        ((if !some_used then check_strict else check) name)
-                  );
-              Env.set_value_used_callback
-                name vd
-                (fun () ->
-                  match !current_slot with
-                  | Some slot ->
-                      slot := (name, vd) :: !slot; rec_needed := true
-                  | None ->
-                      List.iter
-                        (fun (name, vd) -> Env.mark_value_used env name vd)
-                        (get_ref slot);
-                      used := true;
-                      some_used := true
-                )
-            )
-            (Typedtree.pat_bound_idents pat);
-          pat, Some slot
-        )
+    List.map2
+      (fun attrs pat ->
+         Builtin_attributes.warning_scope ~ppwarning:false attrs (fun () ->
+           if not warn_unused then pat, None
```

Comment:
This warn_unused was a fast-path for the case where all relevant warnings are disabled, but now that each binding in the group can have a different configuration, you'd need to define warn_unused as the conjunction of the current warn_unused in the scopes for all bindings.  Or just drop the fast-path altogether (after checking that it does not slow down compilation significantly).

Reply:
To be clear, the proposed code will fail to report warnings is they are globally disabled but re-enabled explicitly for a given binding.

Reply:
Fixed below - as a *disjunction*, right?

Reply:
Well no, for the fast path to be used, you want the warnings to be disabled for *all* the bindings.

Reply:
I think we may be saying the same thing up to De Morgan...
As the fast track is conditioned by `not warn_unused`, you
compute `warn_unused` as the disjunction for all bindings.

Reply:
Ah, `warn_unused` is a bad name...  Could you rename to `warn_used`?

Reply:
Renamed to `warn_about_unused_bindings` (it is lengthy, but used only twice).